src/ua_types_encoding_json.c

```diff
@@ -0,0 +1,5101 @@
+        if (useReversible) {
+            /* REVERSIBLE */
+            ret |= ADDJSONCHAR(ObjStart);
+
+            ret |= calcWriteKey(ctx, "TypeId", UA_FALSE);
+            ret |= CALC_DIRECT(&typeId, NodeId);
+            if (ret != UA_STATUSCODE_GOOD)
+                return ret;
+            
+            /* Encoding field is omitted if the value is 0. */
+            const UA_DataType *contentType = src->content.decoded.type;
+
+            /* Encode the content */
+            ret |= calcWriteKey(ctx, "Body", UA_TRUE);
+            ret |= calcJsonInternal(src->content.decoded.data, contentType, ctx, useReversible);
+            if (ret != UA_STATUSCODE_GOOD)
+                return ret;
+
+            ret |= ADDJSONCHAR(ObjEnd);
+        } else {
+            /* -NON-REVERSIBLE
+             * For the non-reversible form, ExtensionObject values 
+             * shall be encoded as a JSON object containing only the 
+             * value of the Body field. The TypeId and Encoding fields are dropped.
+             * 
+             * TODO: "Body" key in the ExtensionObject?
+             */
+            ret |= ADDJSONCHAR(ObjStart);
+            const UA_DataType *contentType = src->content.decoded.type;
+            ret |= calcWriteKey(ctx, "Body", UA_FALSE);
+            ret |= calcJsonInternal(src->content.decoded.data, contentType, ctx, useReversible);
+            ret |= ADDJSONCHAR(ObjEnd);
+        }
+    }
+    return ret;
+}
+
+static status
+Variant_calcJsonWrapExtensionObject(const UA_Variant *src, const bool isArray, CtxJson *ctx, UA_Boolean useReversible) {
+    size_t length = 1;
+
+    status ret = UA_STATUSCODE_GOOD;
+    if (isArray) {
+        if (src->arrayLength > UA_INT32_MAX)
+            return UA_STATUSCODE_BADENCODINGERROR;
+        
+        length = src->arrayLength;
+    }
+
+    /* Set up the ExtensionObject */
+    UA_ExtensionObject eo;
+    UA_ExtensionObject_init(&eo);
+    eo.encoding = UA_EXTENSIONOBJECT_DECODED;
+    eo.content.decoded.type = src->type;
+    const u16 memSize = src->type->memSize;
+    uintptr_t ptr = (uintptr_t) src->data;
+
+    if(length > 1){
+        ADDJSONCHAR(ArrayStart);
+    }
+    
+    UA_Boolean commaNeeded = false;
+    
+    /* Iterate over the array */
+    for (size_t i = 0; i <  length && ret == UA_STATUSCODE_GOOD; ++i) {
+        if (commaNeeded) {
+            ADDJSONCHAR(Comma);
+        }
+        
+        eo.content.decoded.data = (void*) ptr;
+        ret = calcJsonInternal(&eo, &UA_TYPES[UA_TYPES_EXTENSIONOBJECT], ctx, useReversible);
+        if (ret != UA_STATUSCODE_GOOD)
+            return ret;
+        ptr += memSize;
+        
+        commaNeeded = true;
+    }
+    
+    if(length > 1){
+        ADDJSONCHAR(ArrayEnd);
+    }
+    return ret;
+}
+
+static status
+calcAddMatrixContentJSON(CtxJson *ctx, void* array, const UA_DataType *type, size_t *index, UA_UInt32 *arrayDimensions, size_t dimensionIndex, size_t dimensionSize, UA_Boolean useReversible) {
```

Comment:
Why do we specifically have to have code for matrices?
The binary encoding does not differentiate between vector/matrix/tensor.

Reply:
For me, a matrix always has 2 dimensions.
If you want to handle any number of dimensions, then call this thing multi-array.